autogluon/utils/tabular/ml/models/knn/knn_utils.py

```diff
@@ -0,0 +1,184 @@
+import numpy as np
+from scipy.stats import mode
+from sklearn.utils.extmath import weighted_mode
+import faiss
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+# Rather than try to import non-public sklearn internals, we implement our own weighting functions here
+# These support the same operations as the sklearn functions - at least as far as possible with FAISS
+def _check_weights(weights):
+    """Check to make sure weights are valid"""
+    if weights in (None, 'uniform', 'distance'):
+        return weights
+    elif callable(weights):
+        return weights
+    else:
+        raise ValueError("weights not recognized: should be 'uniform', 'distance', or a callable function")
+
+def _get_weights(dist, weights):
+    """Get the weights from an array of distances and a parameter weights"""
+    if weights in (None, 'uniform'):
+        return None
+    elif weights == 'distance':
+        # if user attempts to classify a point that was zero distance from one
+        # or more training points, those training points are weighted as 1.0
+        # and the other points as 0.0
+        with np.errstate(divide='ignore'):
+            dist = 1. / dist
+        inf_mask = np.isinf(dist)
+        inf_row = np.any(inf_mask, axis=1)
+        dist[inf_row] = inf_mask[inf_row]
+        return dist
+    elif callable(weights):
+        return weights(dist)
+    else:
+        raise ValueError("weights not recognized: should be 'uniform', 'distance', or a callable function")
+
+class KNeighborsRegressor:
+    def __init__(self, n_neighbors = 5, weights='uniform', index_factory_string = "Flat"): 
+        """
+        Creates a KNN regressor model based on FAISS. FAISS allows you to compose different
+        near-neighbor search algorithms from several different preprocessing / search algorithms
+        This composition is specified by the string that is passed to the FAISS index_factory. 
+        Here are good guidelines for choosing the index string: 
+        https://github.com/facebookresearch/faiss/wiki/Guidelines-to-choose-an-index
+
+        The model itself is a clone of the sklearn one
+        """
+        self.index_factory_string = index_factory_string
+        self.n_neighbors = n_neighbors
+        self.weights = weights
+
+
+    def fit(self, X_train, y_train):
+        X_train = X_train.astype(np.float32)
```

Comment:
If we check if X_train is a dataframe, we could do X_train.to_numpy(dtype=np.float32), which may eliminate an unnecessary data copy.

Reply:
Great point - fixed. Also fixed a second unnecessary copy that might happen in the following line.